js-pkg/client/src/provider.ts

```diff
@@ -210,443 +50,392 @@ type WebSocketPolyfillType = {
-  // so we can re-subscribe to them when the provider is re-created
-  const subscribedEvents = new Set<string>()
-
-  const awareness = providerParams.awareness ?? new awarenessProtocol.Awareness(doc)
-  providerParams = { ...providerParams, awareness }
-
-  let _clientToken = initialClientToken ?? (await getClientToken(authEndpoint, roomname))
-  let _provider = new YSweetProvider(_clientToken.url, roomname, doc, {
-    ...updateProviderParams(providerParams, _clientToken),
-  })
-  _provider.addOnFailureHandler(recreateProvider)
-
-  async function recreateProvider() {
-    _clientToken = await getClientToken(authEndpoint, roomname)
-    _provider = new YSweetProvider(_clientToken.url, roomname, doc, {
-      ...updateProviderParams(providerParams, _clientToken),
-      connect: true,
-    })
-    _provider.addOnFailureHandler(recreateProvider)
-    // the previous provider's destroy() method should have been called before
-    // recreateProvider() is called, so we don't need to unsubscribe from events
-    // before re-subscribing to events here
-    for (const event of subscribedEvents) {
-      subscribeToEvent(event)
+    if (this.status.status !== status.status) {
+      this.status = status
+      this.emit('status', status)
     }
   }
 
-  // for each event that is subscribed to on the local observable, make sure to
-  // subscribe to it on the provider
-  function subscribeToEvent(name: string) {
-    _provider.on(name, (...args: any[]) => observable.emit(name, args))
-    subscribedEvents.add(name)
+  private handleAwarenessUpdate(
+    { added, updated, removed }: { added: Array<any>; updated: Array<any>; removed: Array<any> },
+    _origin: any,
+  ) {
+    const changedClients = added.concat(updated).concat(removed)
+    const encoder = encoding.createEncoder()
+    encoding.writeVarUint(encoder, messageAwareness)
+    encoding.writeVarUint8Array(
+      encoder,
+      awarenessProtocol.encodeAwarenessUpdate(this.awareness, changedClients),
+    )
+
+    this.websocket?.send(encoding.toUint8Array(encoder))
+  }
+
+  public destroy() {
+    if (this.websocket) {
+      this.websocket.close()
+    }
+
+    awarenessProtocol.removeAwarenessStates(this.awareness, [this.doc.clientID], 'window unload')
+  }
+
+  /** Aliases for compatibility with y-websocket provider. */
+
+  private _on(type: YSweetEvent, listener: (d: any) => void, once?: boolean): void {
+    if ((type as any) === 'synced') {
+      // Yjs emits both 'sync' and 'synced' events for legacy reasons. We only emit 'sync', but if
+      // we attempt to subscribe to 'synced', we rewrite it to 'sync' to maintain compatibility.
+      // Ref: https://github.com/yjs/y-websocket/blob/aa4220407bda51ab6282d1291de6493c136c2089/src/y-websocket.js#L404
+      type = EVENT_SYNC
+    }
+
+    if (!this.listeners.has(type)) {
+      this.listeners.set(type, new Set())
+    }
+    this.listeners.get(type)?.add(listener)
+    if (once) {
+      this.once(type, listener)
```

Comment:
`this.once` calls `this._on(type, listener, true)` so I think this will recurse infinitely.